executor/pipelined_window.go

```diff
@@ -0,0 +1,459 @@
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	begin, end := e.groupChecker.getNextGroup()
+	e.rowToConsume += uint64(end - begin)
+	for i := begin; i < end; i++ {
+		e.rows = append(e.rows, e.childResult.GetRow(i))
+	}
+	return
+}
+
+func (e *PipelinedWindowExec) fetchChild(ctx context.Context) (EOF bool, err error) {
+	// TODO: reuse chunks
+	childResult := newFirstChunk(e.children[0])
+	err = Next(ctx, e.children[0], childResult)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	// No more data.
+	numRows := childResult.NumRows()
+	if numRows == 0 {
+		return true, nil
+	}
+
+	// TODO: reuse chunks
+	resultChk := chunk.New(e.retFieldTypes, 0, numRows)
+	err = e.copyChk(childResult, resultChk)
+	if err != nil {
+		return false, err
+	}
+	e.accumulated += uint64(numRows)
+	e.data = append(e.data, dataInfo{chk: resultChk, remaining: uint64(numRows), accumulated: e.accumulated})
+
+	e.childResult = childResult
+	return false, nil
+}
+
+func (e *PipelinedWindowExec) copyChk(src, dst *chunk.Chunk) error {
+	columns := e.Schema().Columns[:len(e.Schema().Columns)-e.numWindowFuncs]
+	for i, col := range columns {
+		if err := dst.MakeRefTo(i, src, col.Index); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *PipelinedWindowExec) getRow(i uint64) chunk.Row {
+	return e.rows[i-e.rowStart]
+}
+
+func (e *PipelinedWindowExec) getRows(start, end uint64) []chunk.Row {
+	return e.rows[start-e.rowStart : end-e.rowStart]
+}
+
+// finish is called upon a whole partition is consumed
+func (e *PipelinedWindowExec) finish() {
+	e.whole = true
+}
+
+func (e *PipelinedWindowExec) getStart(ctx sessionctx.Context) (uint64, error) {
+	if e.start.UnBounded {
+		return 0, nil
+	}
+	if e.isRangeFrame {
+		var start uint64
+		start = mathutil.MaxUint64(e.lastStartRow, e.stagedStartRow)
+		for start = e.lastStartRow; start < e.rowCnt; start++ {
+			var res int64
+			var err error
+			for i := range e.orderByCols {
+				res, _, err = e.start.CmpFuncs[i](ctx, e.orderByCols[i], e.start.CalcFuncs[i], e.getRow(start), e.getRow(e.curRowIdx))
+				if err != nil {
+					return 0, err
+				}
+				if res != 0 {
+					break
+				}
+			}
+			// For asc, break when the calculated result is greater than the current value.
+			// For desc, break when the calculated result is less than the current value.
+			if res != e.expectedCmpResult {
+				break
+			}
+		}
+		e.stagedStartRow = start
+		return start, nil
+	}
+	switch e.start.Type {
+	case ast.Preceding:
+		if e.curRowIdx > e.start.Num {
+			return e.curRowIdx - e.start.Num, nil
+		}
+		return 0, nil
+	case ast.Following:
+		return e.curRowIdx + e.start.Num, nil
+	default: // ast.CurrentRow
+		return e.curRowIdx, nil
+	}
+}
+
+func (e *PipelinedWindowExec) getEnd(ctx sessionctx.Context) (uint64, error) {
+	if e.end.UnBounded {
+		return e.rowCnt, nil
+	}
+	if e.isRangeFrame {
+		var end uint64
+		end = mathutil.MaxUint64(e.lastEndRow, e.stagedEndRow)
+		for end = e.lastEndRow; end < e.rowCnt; end++ {
```

Comment:
`for end = mathutil.MaxUint64(e.lastEndRow, e.stagedEndRow); end < e.rowCnt; end++`

Otherwise, what's the means of this modification?

Reply:
sorry for the mistake.